pex/interpreter_constraints.py

```diff
@@ -138,3 +133,71 @@ def create_message(self, preamble=None):
             ),
             problems=problems,
         )
+
+
+class Lifecycle(Enum):
+    class Value(Enum.Value):
+        pass
+
+    DEV = Value("dev")
+    STABLE = Value("stable")
+    EOL = Value("eol")
+
+
+@attr.s(frozen=True)
+class PythonVersion(object):
+    lifecycle = attr.ib()  # type: Lifecycle.Value
+    major = attr.ib()  # type: int
+    minor = attr.ib()  # type: int
+    patch = attr.ib()  # type: int
+
+    def pad(self, padding):
+        # type: (int) -> PythonVersion
+        if self.lifecycle == Lifecycle.EOL:
+            return self
+        return attr.evolve(self, patch=self.patch + padding)
+
+    def iter_compatible_versions(self, specifier_sets):
+        # type: (Iterable[SpecifierSet]) -> Iterator[Tuple[int, int, int]]
+        for patch in range(self.patch + 1):
+            version = (self.major, self.minor, patch)
+            version_string = ".".join(map(str, version))
+            if not specifier_sets:
+                yield version
+            else:
+                for specifier_set in specifier_sets:
+                    if version_string in specifier_set:
+                        yield version
+                        break
+
+
+# TODO(John Sirois): Integrate a `pyenv install -l` based lint / generate script for CI / local
+# use that emits the current max patch for these versions so we automatically stay up to date
+# mod dormancy in the project.
+
+COMPATIBLE_PYTHON_VERSIONS = tuple(
+    # Each 5 units of padding costs `iter_compatible_versions` ~2 extra ms and a padding of 10 gets
+    # us through most of a brand new stable release's likely active lifecycle.
+    version.pad(10)
+    for version in (
+        PythonVersion(Lifecycle.EOL, 2, 7, 18),
```

Comment:
Is that 2 ms per (major,minor) version, or total? If the former, should the padding be per-(major,minor) version? There will never be a 2.7.28. If the latter, there's probably no benefit.

Reply:
Its total for bumping the pad by 5. The 2.7.28 case is already handled since EOL gets no padding. I'll continue the rest of the discussion of this commit under its review in #1575.